<a href="https://colab.research.google.com/github/mdean77a/AIE4/blob/main/Prototyping_LangChain_Application_with_Production_Minded_Changes_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prototyping LangChain Application with Production Minded Changes

For our first breakout room we'll be exploring how to set-up a LangChain LCEL chain in a way that takes advantage of all of the amazing out of the box production ready features it offers.

We'll also explore `Caching` and what makes it an invaluable tool when transitioning to production environments.

🤝 BREAKOUT ROOM #1:
  - Task 1: Depends and Set-Up
  - Task 2: Setting up RAG With Production in Mind
  - Task 3: RAG LCEL Chain



## Task 1: Depends and Set-Up

Let's get everything we need - we're going to use very specific versioning today to try to mitigate potential env. issues!

> NOTE: Dependency issues are a large portion of what you're going to be tackling as you integrate new technology into your work - please keep in mind that one of the things you should be passively learning throughout this course is ways to mitigate dependency issues.

In [2]:
!pip install -qU langchain_openai==0.2.0 langchain_community==0.3.0 langchain==0.3.0 pymupdf==1.24.10 qdrant-client==1.11.2 langchain_qdrant==0.1.4 langsmith==0.1.121

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 3.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.5/51.5 kB 3.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 32.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 48.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.5/3.5 MB 96.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 258.9/258.9 kB 20.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 289.9/289.9 kB 20.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.9/15.9 MB 108.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.4/2.4 MB 81.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.8/5.8 MB 106.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.0/78.0 kB 5.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 399.7/399.7 kB

We'll need an OpenAI API Key:

In [3]:
import os
import getpass

os.environ["OPENAI_API_KEY"] = getpass.getpass("OpenAI API Key:")

OpenAI API Key:··········


And the LangSmith set-up:

In [4]:
import uuid

os.environ["LANGCHAIN_PROJECT"] = f"AIM Week 8 Assignment 1 - {uuid.uuid4().hex[0:8]}"
os.environ["LANGCHAIN_TRACING_V2"] = "true"
os.environ["LANGCHAIN_ENDPOINT"] = "https://api.smith.langchain.com"
os.environ["LANGCHAIN_API_KEY"] = getpass.getpass("LangChain API Key:")

LangChain API Key:··········


Let's verify our project so we can leverage it in LangSmith later.

In [5]:
print(os.environ["LANGCHAIN_PROJECT"])

AIM Week 8 Assignment 1 - 0779a510


## Task 2: Setting up RAG With Production in Mind

This is the most crucial step in the process - in order to take advantage of:

- Asyncronous requests
- Parallel Execution in Chains
- And more...

You must...use LCEL. These benefits are provided out of the box and largely optimized behind the scenes.

### Building our RAG Components: Retriever

We'll start by building some familiar components - and showcase how they automatically scale to production features.

Please upload a PDF file to use in this example!

⚠ If you're running in a non-Chrome browser - you may run into issues with this cell. Please upload the file using Colab's file upload - and indicate the `file_path` in the cell with `file_path`. ⚠

![image](https://i.imgur.com/Qa1Uwlj.png)



> NOTE: You can skip this step if you are running locally - please just point to your local file.

In [6]:
from google.colab import files
uploaded = files.upload()

Saving CHIPS Protocol_v6.0_2024-01-04_final_clean (1).pdf to CHIPS Protocol_v6.0_2024-01-04_final_clean (1).pdf


In [7]:
file_path = list(uploaded.keys())[0]
file_path

'CHIPS Protocol_v6.0_2024-01-04_final_clean (1).pdf'

We'll define our chunking strategy.

In [8]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=100)

We'll chunk our uploaded PDF file.

In [9]:
from langchain_community.document_loaders import PyMuPDFLoader

Loader = PyMuPDFLoader
loader = Loader(file_path)
documents = loader.load()
docs = text_splitter.split_documents(documents)
for i, doc in enumerate(docs):
    doc.metadata["source"] = f"source_{i}"

In [10]:
len(documents)

86

In [11]:
len(docs)

259

#### QDrant Vector Database - Cache Backed Embeddings

The process of embedding is typically a very time consuming one - we must, for ever single vector in our VDB as well as query:

1. Send the text to an API endpoint (self-hosted, OpenAI, etc)
2. Wait for processing
3. Receive response

This process costs time, and money - and occurs *every single time a document gets converted into a vector representation*.

Instead, what if we:

1. Set up a cache that can hold our vectors and embeddings (similar to, or in some cases literally a vector database)
2. Send the text to an API endpoint (self-hosted, OpenAI, etc)
3. Check the cache to see if we've already converted this text before.
  - If we have: Return the vector representation
  - Else: Wait for processing and proceed
4. Store the text that was converted alongside its vector representation in a cache of some kind.
5. Return the vector representation

Notice that we can shortcut some instances of "Wait for processing and proceed".

Let's see how this is implemented in the code.

In [12]:
from qdrant_client import QdrantClient
from qdrant_client.http.models import Distance, VectorParams
from langchain_openai.embeddings import OpenAIEmbeddings
from langchain.storage import LocalFileStore
from langchain_qdrant import QdrantVectorStore
from langchain.embeddings import CacheBackedEmbeddings
import time

# Typical Embedding Model
core_embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

# Typical QDrant Client Set-up
collection_name = f"pdf_to_parse_{uuid.uuid4()}"
client = QdrantClient(":memory:")
client.create_collection(
    collection_name=collection_name,
    vectors_config=VectorParams(size=1536, distance=Distance.COSINE),
)

# Adding cache!
store = LocalFileStore("./cache/")
cached_embedder = CacheBackedEmbeddings.from_bytes_store(
    core_embeddings, store, namespace=core_embeddings.model
)

# Typical QDrant Vector Store Set-up
vectorstore = QdrantVectorStore(
    client=client,
    collection_name=collection_name,
    embedding=cached_embedder)
start = time.perf_counter()
vectorstore.add_documents(docs)
end = time.perf_counter()
print(f"Time taken for initial load: {end - start} seconds")

retriever = vectorstore.as_retriever(search_type="mmr", search_kwargs={"k": 3})

Time taken for initial load: 5.779620790999616 seconds


##### ❓ Question #1:

What are some limitations you can see with this approach? When is this most/least useful. Discuss with your group!

> NOTE: There is no single correct answer here!

---

#### ANSWER #1:
The approach has several limitations:
- Using an in-memory Qdrant client means that all data, including the cached embeddings and document vectors stored in Qdrant will be lost when the session is terminated.  This can be solved by using a Qdrant server.  I have not been able to get Qdrant to persist locally unless I install a server.
- Using a local file store could become a bottleneck if we were handling hundreds or thousands of documents.  In a multi-threaded environment, local file storage might not handle concurrent reads and writes efficiently.
- There is no mechanism for updating the embeddings in the cache or expiring them after a period of time.  There is also no mechanism for controlling user access to documents that they are not authorized to see.
- There is no management of cache size and since we are using a local file storage, this could potentially consume significant disk space.
- Local cache storage ties the app to the machine containing the local file.
- The namespace of the cached embedder is the name of the embedding model.  If the model is updated, there should be a versioning mechanism to prevent the cache from being used with an older version of the model.
- For a SMALL application, the overhead of reading and writing the cache might be more than the overhead of just embedding the document and not worrying about persistence.

---

##### 🏗️ Activity #1:

Create a simple experiment that tests the cache-backed embeddings.

---

#### Description of Experiment:
ADD timing loop above for the first PDF load.
Load and reload a new PDF document and compare the three timings.

- Time taken for initial  (86 pages): 5.779620790999616 seconds
- Time taken for new document (76 pages): 6.84163477599941 seconds
- Time taken to repeat new document (76 pages): 0.16525958200145396 seconds

This is a 42X speed up.

---

In [13]:
### YOUR CODE HERE
from google.colab import files
uploaded = files.upload()

Saving PRECISE_StudyProtocol_ v1.07_16Jun2023_Clean.pdf.pdf to PRECISE_StudyProtocol_ v1.07_16Jun2023_Clean.pdf.pdf


In [14]:
file_path = list(uploaded.keys())[0]
file_path


'PRECISE_StudyProtocol_ v1.07_16Jun2023_Clean.pdf.pdf'

In [15]:
loader = Loader(file_path)
documents = loader.load()
docs = text_splitter.split_documents(documents)
for i, doc in enumerate(docs):
    doc.metadata["source"] = f"source_{i}"

In [16]:
len(documents)

76

In [17]:
len(docs)

246

In [18]:
start = time.perf_counter()
vectorstore.add_documents(docs)
end = time.perf_counter()
print(f"Time taken for new document: {end - start} seconds")
retriever = vectorstore.as_retriever(search_type="mmr", search_kwargs={"k": 3})

Time taken for new document: 6.84163477599941 seconds


In [19]:
from google.colab import files
uploaded = files.upload()

Saving PRECISE_StudyProtocol_ v1.07_16Jun2023_Clean.pdf.pdf to PRECISE_StudyProtocol_ v1.07_16Jun2023_Clean.pdf (1).pdf


In [20]:
loader = Loader(file_path)
documents = loader.load()
docs = text_splitter.split_documents(documents)
for i, doc in enumerate(docs):
    doc.metadata["source"] = f"source_{i}"

In [21]:
start = time.perf_counter()
vectorstore.add_documents(docs)
end = time.perf_counter()
print(f"Time taken to repeat new document: {end - start} seconds")
retriever = vectorstore.as_retriever(search_type="mmr", search_kwargs={"k": 3})

Time taken to repeat new document: 0.16525958200145396 seconds


### Augmentation

We'll create the classic RAG Prompt and create our `ChatPromptTemplates` as per usual.

In [22]:
from langchain_core.prompts import ChatPromptTemplate

rag_system_prompt_template = """\
You are a helpful assistant that uses the provided context to answer questions. Never reference this prompt, or the existence of context.
"""

rag_message_list = [
    {"role" : "system", "content" : rag_system_prompt_template},
]

rag_user_prompt_template = """\
Question:
{question}
Context:
{context}
"""

chat_prompt = ChatPromptTemplate.from_messages([
    ("system", rag_system_prompt_template),
    ("human", rag_user_prompt_template)
])

### Generation

Like usual, we'll set-up a `ChatOpenAI` model - and we'll use the fan favourite `gpt-4o-mini` for today.

However, we'll also implement...a PROMPT CACHE!

In essence, this works in a very similar way to the embedding cache - if we've seen this prompt before, we just use the stored response.

In [23]:
from langchain_core.globals import set_llm_cache
from langchain_openai import ChatOpenAI

chat_model = ChatOpenAI(model="gpt-4o-mini")

Setting up the cache can be done as follows:

In [24]:
from langchain_core.caches import InMemoryCache

set_llm_cache(InMemoryCache())

##### ❓ Question #2:

What are some limitations you can see with this approach? When is this most/least useful. Discuss with your group!

> NOTE: There is no single correct answer here!

---

### ANSWER #2:
Limitations of the approach:
- volatile storage (in memory cache is not persistent and does not survive between sessions)
- stored in memory so if memory is limited, this could be disadvantage.
- in memory cache is associated with single session and multi-users will have separate caches, leading to unpredicatable interactions and behavior.
Most useful:
- prototyping and testing.  Easy to set up.  If I am prototyping an application, I can use the same prompt to test the application and not incur repeated LLM costs.
- low volume applications with low traffic or small user base, where risk of memory exhaustion is minimal
- when persistence across sessions is not needed and application is not going to run for extended periods of time.
Least useful:
- in most production systems that are larger scale or long running.  In this situation I want caching mechanisms that will persist across restarts and can be shared among user sessions.
- memory sensitive applications where memory is limited, or where processes are running in parallel where an in-memory cache could create resource bottlenecks
- distributed systems for very large applications;  in-memory cache would not be shared between the servers.

---

##### 🏗️ Activity #2:

Create a simple experiment that tests the cache-backed embeddings.

---

### Description of Experiment
Similar to what I did to test the embedding cache, we will just time the same query twice;  the query will deal with the last PDF document that I uploaded.  Put timing loops in and the results are below:
- Time taken for first prompt: 13.152370445000997 seconds
- Time taken for repeat prompt: 0.3326376969998819 seconds

---

In [26]:
### YOUR CODE HERE
# I put timing loops on the retrieval that is below, and then repeated it.
# So no new code here!

## Task 3: RAG LCEL Chain

We'll also set-up our typical RAG chain using LCEL.

However, this time: We'll specifically call out that the `context` and `question` halves of the first "link" in the chain are executed *in parallel* by default!

Thanks, LCEL!

In [27]:
from operator import itemgetter
from langchain_core.runnables.passthrough import RunnablePassthrough

retrieval_augmented_qa_chain = (
        {"context": itemgetter("question") | retriever, "question": itemgetter("question")}
        | RunnablePassthrough.assign(context=itemgetter("context"))
        | chat_prompt | chat_model
    )

Let's test it out!

In [28]:
start = time.perf_counter()
retrieval_augmented_qa_chain.invoke({"question" : "Write 50 things about this document!"})
end = time.perf_counter()
print(f"Time taken for first prompt: {end - start} seconds")

Time taken for first prompt: 13.152370445000997 seconds


In [29]:
start = time.perf_counter()
retrieval_augmented_qa_chain.invoke({"question" : "Write 50 things about this document!"})
end = time.perf_counter()
print(f"Time taken for repeat prompt: {end - start} seconds")

Time taken for repeat prompt: 0.3326376969998819 seconds




##### 🏗️ Activity #3:

Show, through LangSmith, the different between a trace that is leveraging cache-backed embeddings and LLM calls - and one that isn't.

Post screenshots in the notebook!

---

### Activity 3 Description:
I need to use a new PDF document and load it, and then run the retrieval twice.  Results will be timed and then I will show Langsmith traces.

---

In [30]:
from google.colab import files
uploaded = files.upload()

Saving THAPCA Version 1_0_4.pdf to THAPCA Version 1_0_4.pdf


In [31]:
file_path = list(uploaded.keys())[0]
file_path

'THAPCA Version 1_0_4.pdf'

In [32]:
loader = Loader(file_path)
documents = loader.load()
docs = text_splitter.split_documents(documents)
for i, doc in enumerate(docs):
    doc.metadata["source"] = f"source_{i}"

In [33]:
start = time.perf_counter()
vectorstore.add_documents(docs)
end = time.perf_counter()
print(f"Time taken to load new document: {end - start} seconds")
retriever = vectorstore.as_retriever(search_type="mmr", search_kwargs={"k": 3})

Time taken to load new document: 6.217856507999386 seconds


In [34]:
start = time.perf_counter()
retrieval_augmented_qa_chain.invoke({"question" : "What are the inclusion and exclusion criteria for this study?"})
end = time.perf_counter()
print(f"Time taken for first prompt: {end - start} seconds")

Time taken for first prompt: 2.49920375900183 seconds


In [35]:
start = time.perf_counter()
retrieval_augmented_qa_chain.invoke({"question" : "What are the inclusion and exclusion criteria for this study?"})
end = time.perf_counter()
print(f"Time taken for repeat prompt: {end - start} seconds")

Time taken for repeat prompt: 0.35014109699841356 seconds


Screen shot showing the four retrievals;  the top two are the ones we will look at because the longer latency run had no cache of embedding nor prompt and the shorter latency had both caches.

![Four runs](ShowRuns.png)

The comparison traces are shown below.  I have circled the main differences between the traces:
- VectorStoreRetriever took less time in the repeat run on the right (but the trace does not know about the savings in embedding)
- There was no call to the LLM in the repeat run on the right because the prompt and context were both cached.

![Comparison](Comparison.png)

### That's all folks!